### ISSUE TITLE
Bug: IndexError in fast_abs_percentile function

### ISSUE THREAD

**Issue by dohmatob** [nilearn/nilearn#872]
```python
import numpy as np
from nilearn._utils.extmath import fast_abs_percentile
fast_abs_percentile(np.arange(4))
```

```
IndexError                                Traceback (most recent call last)
<ipython-input-29-094589b8fe4e> in <module>()
----> 1 fast_abs_percentile(arange(4))

/home/elvis/CODE/FORKED/nilearn/nilearn/_utils/extmath.pyc in fast_abs_percentile(data, percentile)
     43     if partition is not None:
     44         # Partial sort: faster than sort
---> 45         return partition(data, index)[index + 1]
     46     data.sort()
     47     return data[index + 1]

IndexError: index out of bounds
> /home/elvis/CODE/FORKED/nilearn/nilearn/_utils/extmath.py(45)fast_abs_percentile()
     44         # Partial sort: faster than sort
---> 45         return partition(data, index)[index + 1]
     46     data.sort()
```

This bug affects all the `find_cuts` functions of the plotting backend.

**Comment by lesteve** [comment]
@dohmatob I reformatted your comment, next time use fenced blocks delimited by triple backquotes for traceback, otherswise it's pretty hard to read.

**Comment by dohmatob** [comment]
Thx, I updated your comment... :)

**Comment by lesteve** [comment]
OK now I am very confused about fast_abs_percentile, is that expected?

```python

import scipy.stats
from nilearn._utils.extmath import fast_abs_percentile

data = [1, 2, 3, 4]
scipy.stats.scoreatpercentile(data, 50)  # 2.5 

fast_abs_percentile([1, 2, 3, 4], 50)  # 4
```

Also the fact the two branches, the first one using sort and the second one using np.partition seem very unlikely to give the same result unless you use `[index]` and not `[index+1]`. From the documentation of np.partition:

```
Signature: np.partition(a, kth, axis=-1, kind='introselect', order=None)
Docstring:

Return a partitioned copy of an array.

Creates a copy of the array with its elements rearranged in such a way that
the value of the element in kth position is in the position it would be in
a sorted array. All elements smaller than the kth element are moved before
this element and all equal or greater are moved behind it. The ordering of
the elements in the two partitions is undefined.

```

Emphasis mine: **the value of the element in kth position is in the position it would be in
a sorted array**

**Comment by dohmatob** [comment]
This is a fairly valid complain, but I'm afraid it is independent of the PR.

**Comment by lesteve** [comment]
> This is a fairly valid complain, but I'm afraid it is independent of the PR. The PR addresses something else.

Not convinced, my hunch is that we shouldn't use `[index+1]` but `[index]` so that fixes your out-of-bounds indexing too. I.e. both are two different symptoms of the same underlying problem.

**Comment by dohmatob** [comment]
Agreed. I think the partition branch of the code should be simply removed.
All we need do is data.sort() then return data[index + 1].

On Wed, Dec 2, 2015 at 3:48 PM, Loïc Estève <notifications@github.com>
wrote:

> This is a fairly valid complain, but I'm afraid it is independent of the
> PR. The PR addresses something else.
>
> Not convinced, my hunch is that we shouldn't use [index+1] but [index] so
> that fixes your out-of-bounds indexing too. I.e. both are two different
> symptoms of the same underlying problem.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nilearn/nilearn/issues/872#issuecomment-161320551>.
>



-- 
Elvis Dohmatob,
*sent from my  potato...*
http://dohmatob.blogspot.fr/, https://team.inria.fr/parietal/elvis/,
https://github.com/dohmatob

**Comment by dohmatob** [comment]
OK, indeed, better still, we can do the fast_abs_percentile with [index] instead of [index + 1], and remove the sort(...) code snippet. However, note that for real wold data, these threshold will be very close together and the difference is only superficial.

**Comment by lesteve** [comment]
The partition is a speed optimization (np.partition is only available from 1.7 onwards though).

I think that the argument of np.partition should be the same as the one in bracket, because this is the only guarantee given by np.partition as I noted above: "the value of the element in kth position is in the position it would be in a sorted array", and even "All elements smaller than the kth element are moved before this element and all equal or greater are moved behind it. The ordering of
the elements in the two partitions is undefined."

So basically we should have something like that:

```
np.partition(data, k)[k]
```

Then whether to use `index` or `index+1` I am not perfectly sure to be perfectly honest. If it turns out `index+1` is the one then we should definitely do some bound-check as you did in your PR.

**Comment by dohmatob** [comment]
On Wed, Dec 2, 2015 at 4:09 PM, Loïc Estève <notifications@github.com>
wrote:

> The partition is a speed optimization (np.partition is only available from
> 1.7 onwards though).
>
Yes, this is why we do the branching, to begin with.

> I think that the argument of np.partition should be the same as the one in
> bracket, because this is the only guarantee given by np.partition as I
> noted above: "the value of the element in kth position is in the position
> it would be in a sorted array", and even "All elements smaller than the kth
> element are moved before this element and all equal or greater are moved
> behind it. The ordering of
> the elements in the two partitions is undefined."
>
> So basically we should have something like that:
>
> np.partition(data, k)[k]
>
> Then whether to use index or index+1 I am not perfectly sure to be
> perfectly honest. If it turns out index+1 is the one then we should
> definitely do some bound-check as you did in your PR.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nilearn/nilearn/issues/872#issuecomment-161327089>.
>



-- 
Elvis Dohmatob,
*sent from my  potato...*
http://dohmatob.blogspot.fr/, https://team.inria.fr/parietal/elvis/,
https://github.com/dohmatob

**Comment by lesteve** [comment]
> Yes, this is why we do the branching, to begin with.

Yup.

**Comment by GaelVaroquaux** [comment]
> Then whether to use index or index+1 I am not perfectly sure to be perfectly
> honest.

We brainstormed at the sprint, and I can't see a good reason for not
using index. It seems like a simple and good solution.
